presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -488,6 +504,46 @@ public void evaluateFinal(int groupId, BlockBuilder output)
         public void prepareFinal()
         {
         }
+
+        @Override
+        public GroupIdPage addInputPreprocessing(GroupIdPage groupIdPage)
+        {
+            GroupByIdBlock groupIdsBlock = groupIdPage.getGroupByIdBlock();
+
+            // Prepend the groupId block to the input page
+            Page withGroup = groupIdPage.getPage().prependColumn(groupIdsBlock);
+
+            // Filter out positions based on mask, if present
+            Page filtered;
+            if (maskChannel >= 0) {
+                filtered = filter(withGroup, withGroup.getBlock(maskChannel + 1)); // offset by one because of group id in column 0
+            }
+            else {
+                filtered = withGroup;
+            }
+
+            // Compute a mask for the distinct rows (including the group id)
+            Work<Block> work = hash.markDistinctRows(filtered);
+            checkState(work.process());
+            Block distinctMask = work.getResult();
+
+            // Filter out duplicate rows and return the page with distinct rows
+            Page dedupPage = filter(filtered, distinctMask);
```

Comment:
Likely we could do a `compact` when filtering ratio is high.

Reply:
I think even if it compacted every time it would still be a 2x memory regression because it is storing the distinct values twice compared to the non-spill case where the values are stored once in `MarkDistinctHash` and 0-1x inside the wrapped accumulator.

Reply:
I am currently working on adding support for using MarkDistinctHash as the input source when pre-processing is enabled. I have the changes ready and unit tests are passing but I want to do some verifier runs to ensure that changes are working correctly. With these changes, we will no longer buffer pages in `rawInputs` when pre-processing is enabled. Instead, we will get them from `markDistinctHash` right before spilling to disk or pushing to underlying delegate.

With these changes, I am going to restructure the code in the following way:
1. `SpillableFinalOnlyGroupedAccumulator` will be used when pre-processing is disabled or when delegate is OrderingGroupedAccumulator. Some changes are done in this class to support pre-processed data but these changes are backward compatible with non-processed inputs as well.
2. We will add a new class called `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator` which extends `SpillableFinalOnlyGroupedAccumulator` and is used only for distinct aggregates when pre-processing is enabled. This will allow us to write distinct specific logic without adding too many methods in `FinalOnlyGroupedAccumulator` interface.
3. We will remove the `addInputPreprocessing()` and `reset()` API from `FinalOnlyGroupedAccumulator` since these applies only to distinct aggregates and could be called directly inside `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator`
4. `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator` will override necessary methods like addInput, evaluateIntermediate, prepareFinal and will perform logic that is specific to distinct aggregates with preprocessing enabled.

With these changes, we will be able to cleanly separate out logic when pre-processing is enabled vs disabled. I will update the PR once I have done some verifier testing but would be eager to get feedback on the new design.